src/xtgeo/grid3d/_grdecl_grid.py

```diff
@@ -0,0 +1,644 @@
+        zcorn = self.zcorn.reshape((2, nx, 2, ny, 2, nz), order="F")
+
+        if not np.allclose(
+            zcorn[:, :, :, :, 1, : nz - 1], zcorn[:, :, :, :, 0, 1:], atol=1e-2
+        ):
+
+            raise ValueError("xtgeo does not support grids with horizontal split.")
+        result = np.zeros((nx + 1, ny + 1, nz + 1, 4), dtype=np.float32)
+
+        # xtgeo uses 4 z values per i,j,k to mean the 4 z values of
+        # adjacent cells for the cornerline at position i,j,k assuming
+        # no difference in z values between upper and lower cells. In
+        # the order sw,se,nw,ne.
+
+        # In grdecl, there are 8 zvalues per i,j,k meaning the z values
+        # of each corner for the cell at i,j,k. In
+        # the order "left" (west) before "right" (east) , "near" (south)
+        # before "far" (north) , "upper" before "bottom"
+
+        # set the nw value of cornerline i+1,j to
+        # the near right corner of cell i,j
+        result[1:, :ny, 0:nz, 2] = zcorn[1, :, 0, :, 0, :]
+        result[1:, :ny, nz, 2] = zcorn[1, :, 0, :, 1, nz - 1]
+
+        # set the ne value of cornerline i,j to
+        # the near left corner of cell i,j
+        result[:nx, :ny, 0:nz, 3] = zcorn[0, :, 0, :, 0, :]
+        result[:nx, :ny, nz, 3] = zcorn[0, :, 0, :, 1, nz - 1]
+
+        # set the sw value of cornerline i+1,j+1 to
+        # the far right corner of cell i,j to
+        result[1:, 1:, 0:nz, 0] = zcorn[1, :, 1, :, 0, :]
+        result[1:, 1:, nz, 0] = zcorn[1, :, 1, :, 1, nz - 1]
+
+        # set the se value of cornerline i,j+1 to
+        # the far left corner of cell i,j
+        result[:nx, 1:, 0:nz, 1] = zcorn[0, :, 1, :, 0, :]
+        result[:nx, 1:, nz, 1] = zcorn[0, :, 1, :, 1, nz - 1]
+
+        self.duplicate_insignificant_xtgeo_zcorn(result)
+
+        return np.ascontiguousarray(result)
+
+    def transform_xtgeo_coord_by_mapaxes(self, coord):
+        """Transforms xtgeo coord values by mapaxes.
+
+        The mapaxes keyword in a grdecl file defines a new coordinate system by
+        which x and y values are to be interpreted. The given xtgeo coord
+        values are transformed from the local coordinate system defined by
+        mapaxes to global coordinates.
+        """
+        x_point = self.mapaxes.x_line
+        y_point = self.mapaxes.y_line
+        origin = self.mapaxes.origin
+
+        x_axis = np.array(x_point) - origin
+        y_axis = np.array(y_point) - origin
+
+        x_unit = x_axis / np.linalg.norm(x_axis)
+        y_unit = y_axis / np.linalg.norm(y_axis)
+
+        coord[:, :, (0, 1)] = (
+            origin
+            + coord[:, :, 0, np.newaxis] * x_unit
+            + coord[:, :, 1, np.newaxis] * y_unit
+        )
+        coord[:, :, (3, 4)] = (
+            origin
+            + coord[:, :, 3, np.newaxis] * x_unit
+            + coord[:, :, 4, np.newaxis] * y_unit
+        )
+
+        return coord
+
+    def duplicate_insignificant_xtgeo_zcorn(self, zcorn):
```

Comment:
This is never tested directly, only indirectly. Although what it does is not that complicated, is quite easy to get it wrong, could you add a couple of simple tests?

Reply:
I'll add some tests, it was tested against the old implementation before.

Reply:
I added a simplified test for this, but it does not cover all nuances of the duplication, but can be understood at a glance.

Reply:
I added two simplified tests for this, but it does not cover all nuances of the duplication, but can be understood at a glance.